snapshot - index out of range

```
DEBUG[07-29|08:02:46] Snapshot process takes                   time=72.412µs  hash=cf7a02…306354
panic: runtime error: index out of range

goroutine 5832 [running]:
github.com/ethereum/go-ethereum/tomox.(*Tree).Keys(0xc004030140, 0xc005250000, 0xc007bb5230, 0x2f, 0x2f)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/redblacktree.go:210 +0x112
github.com/ethereum/go-ethereum/tomox.prepareOrderTreeData(0xc003d12780, 0xc00729d7e0, 0x0, 0x0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/snapshot.go:111 +0x127
github.com/ethereum/go-ethereum/tomox.newSnapshot(0xc00011b440, 0xf83aeb1c3f027acf, 0xee476b1c381df70f, 0x5000b24a39f38771, 0x546330993d68de54, 0x31466d855f, 0x2337380, 0x546330993d68de54)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/snapshot.go:83 +0x1cd
github.com/ethereum/go-ethereum/tomox.(*TomoX).Snapshot(0xc00011b440, 0xf83aeb1c3f027acf, 0xee476b1c381df70f, 0x5000b24a39f38771, 0x546330993d68de54, 0x0, 0x0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/tomox.go:1179 +0x122
github.com/ethereum/go-ethereum/core.(*BlockChain).insertBlock(0xc000466000, 0xc00b0721b0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/blockchain.go:1447 +0x1d78
github.com/ethereum/go-ethereum/core.(*BlockChain).InsertBlock(0xc000466000, 0xc00b0721b0, 0x546330993d68de54, 0x0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/blockchain.go:1275 +0x39
github.com/ethereum/go-ethereum/eth.NewProtocolManager.func6(0xc00b0721b0, 0x0, 0x0)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/handler.go:180 +0x342
github.com/ethereum/go-ethereum/eth/fetcher.(*Fetcher).insert.func1(0xc00032e1e0, 0x6e23d24835af878c, 0xd2a52c959d59ce0d, 0xa6006916cbc8f9f8, 0x488c879fb61d25a2, 0xc007858180, 0xc002d81710, 0x10)
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/fetcher/fetcher.go:713 +0x1566
created by github.com/ethereum/go-ethereum/eth/fetcher.(*Fetcher).insert
	/root/gocode/tomo/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/fetcher/fetcher.go:655 +0x2f3```

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -206,7 +206,7 @@ func (tree *Tree) Size() uint64 {
 func (tree *Tree) Keys(dryrun bool) [][]byte {
 	keys := make([][]byte, tree.size)
 	it := tree.Iterator()
-	for i := 0; it.Next(dryrun); i++ {
+	for i := 0; it.Next(dryrun) && i < len(keys); i++ {
 		keys[i] = it.Key()
 	}
 	return keys
@@ -216,7 +216,7 @@ func (tree *Tree) Keys(dryrun bool) [][]byte {
 func (tree *Tree) Values(dryrun bool) [][]byte {
 	values := make([][]byte, tree.size)
 	it := tree.Iterator()
-	for i := 0; it.Next(dryrun); i++ {
+	for i := 0; it.Next(dryrun) && i < len(values); i++ {
 		values[i] = it.Value()
 	}
 	return values

--- a/tomox/snapshot.go
+++ b/tomox/snapshot.go
@@ -104,9 +104,10 @@ func prepareOrderTreeData(tree *OrderTree) (*OrderTreeSnapshot, error) {
 	snap.OrderTreeItem = serializedTree
 
 	snap.OrderList = make(map[common.Hash]*OrderListSnapshot)
-	if !tree.NotEmpty() {
+	if !tree.NotEmpty() || tree.PriceTree.Size() == 0 {
 		return snap, nil
 	}
+
 	// foreach each price, snapshot its orderlist
 	for _, key := range tree.PriceTree.Keys(false) {
 		priceKeyHash := common.BytesToHash(key)

--- a/tomox/tomox.go
+++ b/tomox/tomox.go
@@ -1176,7 +1176,7 @@ func (tomox *TomoX) Snapshot(blockHash common.Hash) error {
 	}(time.Now())
 
 	if snap, err = newSnapshot(tomox, blockHash); err != nil {
-		return nil
+		return err
 	}
 	if err = snap.store(tomox.db); err != nil {
 		return err

